src/vos/vos_ilog.c

```diff
@@ -434,8 +434,8 @@ int vos_ilog_update_(struct vos_container *cont, struct ilog_df *ilog,
 		return rc;
 	}
 
-	rc = ilog_update(loh, &max_epr, epr->epr_hi, dtx_is_valid_handle(dth) ?
-			 dth->dth_op_seq : VOS_SUB_OP_MAX, false);
+	rc = ilog_update(loh, &max_epr, epr->epr_hi,
+			 (dtx_is_real_handle(dth) ? dth->dth_op_seq : VOS_SUB_OP_MAX), false);
```

Comment:
As my understand, one local TX may contain multiple sub-modifications, these sub-modifications belong to the same transaction, shares the same major epoch, and different minor epoch for each one. Here, the 4th parameter is "minor_eph". If we use "VOS_SUB_OP_MAX" for local TX case without distinguishing sub-modifications, then it will be explained as either "conflict" or "already done". That does no match the semantics.

Reply:
@janekmi, I discussed this issue with Fan Yong. We speculate the difficult you face here may be that `rdb` does not increment `dth_op_seq` as `obj` does (via `dtx_sub_init`). If that is the case, would this `rdb` change (https://github.com/daos-stack/daos/pull/12330/commits/961c63a273e567353220c359e8dd24c1be115256) allow you to use `dth_op_seq` in the same manner as the non-local TX case?

Reply:
In the [latest fix](https://github.com/daos-stack/daos/pull/13495/commits/3c72881125bcf601659b0916c5daa4f0295f89b0) to this pull request, I introduced both increasing `dth_op_seq` on each update and punch (a little bit different semantics compared to calling `dtx_sub_init()` but should do the trick) and using `dth_op_seq` instead of `VOS_SUB_OP_MAX` exactly as @Nasf-Fan suggested.

Moreover, I wrote additional unit tests ((here)[https://github.com/daos-stack/daos/pull/13495/commits/7592898e87746d80f985df7672a557f7560f5b9a]) after consulting with @Nasf-Fan to verify whether the introduced changes really solved the issue with the conflicting operations. The test with conflicting update and punch ((here)[https://github.com/daos-stack/daos/pull/13495/commits/7592898e87746d80f985df7672a557f7560f5b9a#diff-915f0cd50d1f423d5f44bcc9b8ad81d0e6589df2f932b1cd26f696fe179fdd2bR437]) is currently failing.

From what I have learned ILOG does NOT consider the minor epoch as long as it does not consider the conflicting operations as belonging to the same transaction. Since the local transaction as of now does not track ILOG modifications belonging to a transaction all ILOG modifications are immediately considered `DTX_LID_COMMITTED`. Hence, each operation is considered to belong to different transactions and returns with `DER_TX_RESTART` on conflict. No minor epochs are even considered.

> We speculate the difficult you face here may be that rdb does not increment dth_op_seq as obj does (via dtx_sub_init).

No. This won't solve the problem.

I consulted @jolivier23 and if I cite him correctly, the only way out of this conflict is by introducing a special case just for local transactions to ignore this conflict.

Reply:
@Nasf-Fan, sounds like our understanding was wrong.

@janekmi, would "every VOS local TX must have valid TX ID" solve the problem?

Reply:
> would "every VOS local TX must have valid TX ID" solve the problem?

Sadly no. To track modifications to ILOGs done by the transaction would require adding to the local transaction a DTX entry as a regular DTX transaction has. The local ID of the DTX entry is what allows to differentiate between transactions. Note it would increase the complexity of a local transaction since all ILOG records would have to be updated as committed when the transaction concludes.

Reply:
@janekmi , would you please to apply the following patch for such TX_RESTAR issue?

diff --git a/src/vos/vos_ilog.c b/src/vos/vos_ilog.c
index 065f6ed816..1e06e3452b 100644
--- a/src/vos/vos_ilog.c
+++ b/src/vos/vos_ilog.c
@@ -57,7 +57,7 @@ vos_ilog_is_same_tx(struct umem_instance *umm, uint32_t tx_id,
                 * transactional, treat it as the same transaction and let the
                 * minor epoch handle any conflicts.
                 */
-               if (!dtx_is_valid_handle(dth))
+               if (!dtx_is_real_handle(dth))
                        *same = true;
        } else if (tx_id == dtx) {
                *same = true;

Reply:
@Nasf-Fan:

1. Are we 100% sure RDB is ok with ignoring this kind of conflict? @liw 
2. I believe, this solution is not fully correct. Since it allows comparing minor epochs between transactions whereas the minor epoch scope seems to be correct only in the transaction scope. Am I missing something?

Reply:
It is the same result as without your patch, right?
If with your patch, the local transaction with the same major epoch, should be regarded as belong to the same RDB transaction, it is expected, right? The left things will be handled via minor epoch subsequently, please check update_inplace() logic.

I cannot say it is 100% correct, but please verify whether it can resolve your current test failure. If it did not work, then we will consider other solution.

Reply:
> 2. I believe, this solution is not fully correct. Since it allows comparing minor epochs between transactions whereas the minor epoch seems to be valid only in the transaction scope. Am I missing something?

I got a clarification from @jolivier23. Minor epochs can be compared between transactions since each of the transactions is expected to use a different epoch. Assuming it is ok with RDB works (@liw?) your solution @Nasf-Fan is ok. 👍

Reply:
As my understand, for a RDB transaction, it may contain several sub-modifications, each of them will have different minor epochs but with the same major epoch. That is basically the same as distributed transaction does.

Reply:
> As my understand, for a RDB transaction, it may contain several sub-modifications, each of them will have different minor epochs but with the same major epoch. That is basically the same as distributed transaction does.

Sorry, @Nasf-Fan, I forgot about the metadata container (MC), although the conclusion should be the same. While what you said is true for the log container (LC), all writes (from all `vos` TXs) to the MC use the hard-coded epoch: 1. But all these writes are updates; there's no punches. So sounds like we should be okay.

Reply:
> > As my understand, for a RDB transaction, it may contain several sub-modifications, each of them will have different minor epochs but with the same major epoch. That is basically the same as distributed transaction does.
> 
> Sorry, @Nasf-Fan, I forgot about the metadata container (MC), although the conclusion should be the same. While what you said is true for the log container (LC), all writes (from all `vos` TXs) to the MC use the hard-coded epoch: 1. But all these writes are updates; there's no punches. So sounds like we should be okay.

"hard coded epoch: 1" means for different RDB transactions? If yes, how to distinguish two RDB transactions?

Reply:
All writes to the MC use hard-coded epoch 1---all such writes, from all `rdb` TXs, in all DAOS systems.

Reply:
I don't understand the question "how to distinguish two RDB transactions?". Distinguish for what purposes?

Reply:
Then there will be single RDB transaction? All RDB updates belong to the unique RDB transaction?

Reply:
In my current plan, `rdb` will perform each of its `vos` TXs like this:

```
dtx_begin(..., &dth);

/* Modify the LC. */
vos_obj_update(LC object, ..., epoch=X, ..., &dth);
vos_obj_punch(LC object, ..., epoch=X, &dth);
...

/* Modify the MC (to record X). */
vos_obj_update(MC object, ..., epoch=1, ..., &dth);

dtx_end(&dth, ...);
```

The set of sub-modification to the LC is determined by `rdb` users. The MC modification is always one update to the same <object, dkey, akey> and always in epoch 1.

Reply:
> In my current plan, `rdb` will perform each of its `vos` TXs like this:
> 
> ```
> dtx_begin(..., &dth);
> 
> /* Modify the LC. */
> vos_obj_update(LC object, ..., epoch=X, ..., &dth);
> vos_obj_punch(LC object, ..., epoch=X, &dth);
> ...
> 
> /* Modify the MC (to record X). */
> vos_obj_update(MC object, ..., epoch=1, ..., &dth);
> 
> dtx_end(&dth, ...);
> ```
> 
> The set of sub-modification to the LC is determined by `rdb` users. The MC modification is always one update to the same <object, dkey, akey> and always in epoch 1.

Then why not use the same epoch "X" for MC sub-modification? In your current plan, the sub-modifications in one RDB transaction will have different major epochs. That breaks current VOS TX assumption: the sub-modifications belong to the same transaction will use the same major epoch. In your case, the destination for LC and MC sub-modifications are different, that can work. But consider next RDB transaction, its MC sub-modification will also use the fixed major epoch 1, then ilog logic will regard it as belong to former RDB TX, that is confused, may cause trouble, need to be verified.

Reply:
Because "always using epoch 1 for MC writes" is what `rdb` has been doing since it started to use `vos`. It is part of `rdb`'s storage layout definition. If possible, I don't want to make any changes to that at this time.

Reply:
> That breaks current VOS TX assumption: the sub-modifications belong to the same transaction will use the same major epoch.

IIRC, that is a DTX assumption, not a VOS local TX assumption. For instance, a umem/pmem TX has no "epoch" or "version" concept in its API.

Reply:
> Because "always using epoch 1 for MC writes" is what `rdb` has been doing since it started to use `vos`. It is part of `rdb`'s storage layout definition. If possible, I don't want to make any changes to that at this time.

I cannot understand how ilog logic can distinguish different RDB transactions to use the same epoch and handle them as different transactions. Is it possible the subsequent updates to be skipped silently because of the "same" check? not sure, please verify. But I do not think that is a strong reason to have to use fixed epoch. What is the trouble if MC write uses the same epoch X as LC write does? I may miss some important factor.


> > That breaks current VOS TX assumption: the sub-modifications belong to the same transaction will use the same major epoch.
> 
> IIUC, that is a DTX assumption, not a VOS local TX assumption. For instance, a umem/pmem TX has no "epoch" or "version" concept in its API.

It is current ilog logic determined, not DTX assumption. These part changes do not touch non-local DTX logic. Please check the logic check_equal(), it firstly checks the epoch of the modification. From ilog logic perspective, major epoch is the basic clew to distinguish the transactions unless we (define and) offer more identifier. But I do not think that will be more simpler than flexible epoch for different transactions.
As my understand, VOS is version based, epoch is the version, it should be at least locally distinguishable.

Reply:
> Is it possible the subsequent updates to be skipped silently because of the "same" check? not sure, please verify.

I have verified that the update at the hard-coded epoch 1 works:

```
01/29-05:55:13.30 boro-66 DAOS[487731/0/1959] rdb  DBUG src/rdb/rdb_internal.h:394 rdb_mc_lookup() mc=7fb9d5615650 oid=1 akey=<0x7fb9e77fd9f5, 3> value=<0x7fb9db624270, 304, 304>
01/29-05:55:13.30 boro-66 DAOS[487731/0/1959] rdb  DBUG src/rdb/rdb_raft.c:1211 rdb_raft_log_offer_single() b276ec5f[0]: before update: tail=2
01/29-05:55:13.30 boro-66 DAOS[487731/0/1959] rdb  DBUG src/rdb/rdb_internal.h:382 rdb_mc_update() mc=7fb9d5615650 oid=1 n=1 akeys[0]=<0x7fb9e77fd9f5, 3> values[0]=<0x7fb9d5875280, 304>
01/29-05:55:13.30 boro-66 DAOS[487731/0/1959] rdb  DBUG src/rdb/rdb_internal.h:394 rdb_mc_lookup() mc=7fb9d5615650 oid=1 akey=<0x7fb9e77fd9f5, 3> value=<0x7fb9db624270, 304, 304>
```

An `rdb_mc_lookup` call was added before and after an `rdb_mc_update` call, to read the value before and after the update. All the two `lookup` calls and the `update` call used hard-coded epoch 1.

> But I do not think that is a strong reason to have to use fixed epoch. What is the trouble if MC write uses the same epoch X as LC write does? I may miss some important factor.

The important thing is that "using hard-coded epoch 1" is part of the `rdb` storage layout definition. Changing it would introduce compatibility concerns. Secondly, while I'm okay with using monotonically increasing epochs for MC writes, I'm against using _the same epoch_ as the LC writes, not because of any problem, but because of the confusing model.

Reply:
> > Is it possible the subsequent updates to be skipped silently because of the "same" check? not sure, please verify.
> 
> I have verified that the update at the hard-coded epoch 1 works:
> 
> ```
> 01/29-05:55:13.30 boro-66 DAOS[487731/0/1959] rdb  DBUG src/rdb/rdb_internal.h:394 rdb_mc_lookup() mc=7fb9d5615650 oid=1 akey=<0x7fb9e77fd9f5, 3> value=<0x7fb9db624270, 304, 304>
> 01/29-05:55:13.30 boro-66 DAOS[487731/0/1959] rdb  DBUG src/rdb/rdb_raft.c:1211 rdb_raft_log_offer_single() b276ec5f[0]: before update: tail=2
> 01/29-05:55:13.30 boro-66 DAOS[487731/0/1959] rdb  DBUG src/rdb/rdb_internal.h:382 rdb_mc_update() mc=7fb9d5615650 oid=1 n=1 akeys[0]=<0x7fb9e77fd9f5, 3> values[0]=<0x7fb9d5875280, 304>
> 01/29-05:55:13.30 boro-66 DAOS[487731/0/1959] rdb  DBUG src/rdb/rdb_internal.h:394 rdb_mc_lookup() mc=7fb9d5615650 oid=1 akey=<0x7fb9e77fd9f5, 3> value=<0x7fb9db624270, 304, 304>
> 01/29-05:55:13.30 boro-66 DAOS[487731/0/1959] rdb  DBUG src/rdb/rdb_raft.c:1228 rdb_raft_log_offer_single() b276ec5f[0]: after update: tail=3
> ```
> 

According to current master logic, if two RDB transactions use the same fixed (major) epoch 1, then ilog logic will only do the first update, and commit. When the second RDB transaction arrives, it changes nothing for ilog (obj/key), and only updates the value (SV). But I am wondering how the following logic can be passed:

```
static int
svt_rec_update(struct btr_instance *tins, struct btr_record *rec,
                d_iov_t *key_iov, d_iov_t *val_iov, d_iov_t *val_out)
{
        struct vos_svt_key      *skey;
        struct vos_irec_df      *irec;
        struct vos_rec_bundle   *rbund;
        int                      rc;

        rbund = iov2rec_bundle(val_iov);
        skey = (struct vos_svt_key *)key_iov->iov_buf;
        irec = vos_rec2irec(tins, rec);

        /** Disallow same epoch overwrite */
        if (skey->sk_minor_epc <= irec->ir_minor_epc)
                return -DER_NO_PERM;
```

Unless each LC update use different object/key. My understand the RDB transactions uses single value update.

Reply:
Could the following lines from `btr_update` provide an answer?

```
	rc = btr_rec_update(tcx, rec, key, val, val_out);
	if (rc == -DER_NO_PERM) { /* cannot make inplace change */
		struct btr_trace *trace = &tcx->tc_trace.ti_trace[tcx->tc_depth - 1];

		if (btr_has_tx(tcx)) {
			rc = btr_node_tx_add(tcx, trace->tr_node);
			if (rc != 0)
				goto out;
		}

		D_DEBUG(DB_TRACE, "Replace the original record\n");
		rc = btr_rec_free(tcx, rec, NULL);
		if (rc)
			goto out;
		rc = btr_rec_alloc(tcx, key, val, rec, val_out);
	}
```

Reply:
Yes, it seems so. Based on current ilog implementation, it will not forbid two RDB transactions to use the same fixed epoch 1. With @janekmi 's patch, if the subsequent RDB transaction has more sub-modifications than former ones, it may generate more ilog entries or related object/key, not sure whether it is RDB expected or not. The other parts are basically fine for me.

Reply:
With this PR, it *seems* ilog entries were not generated for repeated updates in the MC during my testing. (They were generated for LC writes, but that was expected, I think.)